flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageMemoryManagerImpl.java

```diff
@@ -0,0 +1,279 @@
+                                    .setUncaughtExceptionHandler(FatalExitExceptionHandler.INSTANCE)
+                                    .build());
+        }
+
+        this.isInitialized = true;
+    }
+
+    @Override
+    public void listenBufferReclaimRequest(Runnable onBufferReclaimRequest) {
+        bufferReclaimRequestListeners.add(onBufferReclaimRequest);
+    }
+
+    /**
+     * Request a {@link BufferBuilder} instance from {@link BufferPool} for a specific owner. The
+     * {@link TieredStorageMemoryManagerImpl} will not check whether a buffer can be requested and
+     * only record the total number of requested buffers. If the buffers in the {@link BufferPool}
+     * is not enough, this will request each tiered storage to reclaim the buffers as much as
+     * possible.
+     *
+     * <p>Note that synchronization is not necessary for this method, as it can only be called by
+     * the task thread.
+     *
+     * @return the requested buffer
+     */
+    @Override
+    public BufferBuilder requestBufferBlocking() {
+        checkIsInitialized();
+
+        tryReclaimBuffers();
+        CompletableFuture<MemorySegment> requestedBuffer =
+                CompletableFuture.supplyAsync(
+                        () -> {
+                            try {
+                                return bufferPool.requestMemorySegmentBlocking();
+                            } catch (InterruptedException e) {
+                                throw new RuntimeException(e);
+                            }
+                        });
+
+        ScheduledFuture<?> scheduledFuture = getPeriodicalCheckReclaimBufferFuture();
+
+        MemorySegment memorySegment = null;
+        try {
+            memorySegment = requestedBuffer.get();
+        } catch (InterruptedException | ExecutionException e) {
+            ExceptionUtils.rethrow(e);
+        }
+
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(false);
+        }
+
+        numRequestedBuffers.incrementAndGet();
+        return new BufferBuilder(checkNotNull(memorySegment), this::recycleBuffer);
+    }
+
+    /**
+     * The synchronization is not necessary for this method, as it can only be called by the task
+     * thread.
+     *
+     * <p>The accuracy of the memory usage ratio may be compromised due to the varying buffer pool
+     * sizes. However, this only impacts a single iteration of the buffer usage check. Upon the next
+     * iteration, the buffer reclaim will eventually be triggered.
+     */
+    @Override
+    public int getMaxNonReclaimableBuffers(Object owner) {
+        checkIsInitialized();
+
+        TieredStorageMemorySpec ownerMemorySpec = checkNotNull(tieredMemorySpecs.get(owner));
+        int ownerGuaranteedBuffers = ownerMemorySpec.getNumGuaranteedBuffers();
+        return bufferPool.getNumBuffers()
+                - numRequestedBuffers.get()
+                - numTotalGuaranteedBuffers
+                + ownerGuaranteedBuffers;
```

Comment:
The calculation doesn't seem right.
- This implies all the requested buffers are non-guaranteed.
- The calculation result can be less than the owner's guaranteed buffer, or even negative.

Reply:
Refactoring it by returning `bufferPool.getNumBuffers() - numBuffersOfOtherOwners`. 

The buffers of other owner (except for the request owner itself) is calculated by `Math.max(numGuaranteed, numRequested)`.